state/protocol/protocol_state.go

```diff
@@ -83,14 +83,16 @@ type ProtocolState interface {
 	GlobalParams() GlobalParams
 }
 
+// MutableProtocolState is the read-write interface for protocol state, it allows evolving the protocol state by
+// creating a StateMutator for each block and applying state-changing service events.
 type MutableProtocolState interface {
 	ProtocolState
 
 	// Mutator instantiates a `StateMutator` based on the previous protocol state.
 	// Has to be called for each block to evolve the protocol state.
 	// Expected errors during normal operations:
 	//  * `storage.ErrNotFound` if no protocol state for parent block is known.
-	Mutator(candidateView uint64, parentID flow.Identifier) (StateMutator, error)
+	Mutator(candidate *flow.Header) (StateMutator, error)
```

Comment:
### ⚠️  

I strongly disagree with this change. There are two distinct scenarios that the Mutator has to satisfy:
1. Checking another node's block proposal. Here, we have a candidate block. All good. 
2. Constructing our own block proposal: 

   Here, we do **_not_ have a candidate header**! The header contains the payload's hash and the payload contains the commitment to the output state. We instantiate the `Mutator` to compute the output state. Therefore, by causality the primary cannot have a valid block header. 

In conclusion, a proper `flow.Header` does not exist for all application scenarios of this method.

Reply:
_Further Context:_ 

I am aware that the current way of building our blocks has this notion of "header under construction" and the type of this header is unfortunately also `flow.Header`.  I consider this _very_ bad technical debt that urgently needs to be resolved (though not part of this work stream). There are a bunch of really nasty complexities resulting from this tech debt:
* You can ask a "header under construction" for its ID. It will give you one, but that is not the ID of the final block, because some fields have dummy values.  This is an invitation for subtle bugs. If I am not mistaken, you stepped right into the trap already in your own PR [here](https://github.com/onflow/flow-go/blob/d69920a457a1dbde4c2a01485414d8645a1b810f/state/protocol/protocol_state/state/mutator.go#L106-L108).
* On a broader scale of the code base, exactly this non-existent differentiation between `entity under construction` and the `final entity` is what is causing all the malleability edge-cases in our code base: 
   * Engineer A adds field `x` to the entity but explicitly excludes it from the entity's ID computation, because the entity is still under construction, and he wants to refer to the entity with the same ID despite still needing to add that one last field. 
   * Engineer B (or Engineer A after some months) sees the field `x` in the entity and uses it for something else. Though, as `x` is not covered by the hash computation, byzantine actors can manipulate the value at will and continuously. Furthermore, they can use different values of `x` to undermine the networking layer's deduplication logic to spam the application layer. 

Therefore, we will (hopefully this year) add a clean separation by types for "entity under construction" vs "final entity". For the primary, you will simply only have the "header under construction" when initializing the `Mutator` which cannot give you a blockID to index the protocol state with!

Reply:
Suggestion: 
 * keep the previous interface
 * you cannot move the indexing logic into the Mutator for causality reasons